File: mobile/welcome-screen/ios-app/ShowUpAPI/Sources/ShowUpAPI/TokenStore.swift
import Foundation

/// Where the access and refresh tokens live, and the only thing the network layer knows about auth.
///
/// A protocol so the middleware depends on behaviour rather than on the Keychain: that is what lets
/// the auth layer be tested with no entitlements, no signed app and no device, using
/// ``InMemoryTokenStore``.
///
/// `Sendable` because the middleware is handed across concurrency domains with the store captured.
public protocol TokenStoring: Sendable {
    /// The current access token, or nil when nobody is signed in.
    func accessToken() async -> String?

    /// The current refresh token, or nil when nobody is signed in.
    func refreshToken() async -> String?

    /// Replace both tokens. Called after sign-in and after every successful refresh.
    func save(accessToken: String, refreshToken: String) async

    /// Forget both tokens: sign-out, and also a failed refresh -- at that point the session is
    /// over whatever the user does next, and a dead token only produces a second confusing failure.
    func clear() async
}

/// A token store held in memory.
///
/// An `actor` rather than a class, so it is safe to share without `@unchecked Sendable` -- which
/// this project forbids and `audit/check-swift-concurrency.py` fails the build on.
///
/// Used by the tests, and deliberately not by the app: tokens must survive the process being
/// killed, which is what ``KeychainTokenStore`` is for.
public actor InMemoryTokenStore: TokenStoring {
    private var access: String?
    private var refresh: String?

    public init(accessToken: String? = nil, refreshToken: String? = nil) {
        self.access = accessToken
        self.refresh = refreshToken
    }

    public func accessToken() async -> String? { access }
    public func refreshToken() async -> String? { refresh }

    public func save(accessToken: String, refreshToken: String) async {
        access = accessToken
        refresh = refreshToken
    }

    public func clear() async {
        access = nil
        refresh = nil
    }
}

File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/api/TokenStore.kt
package com.showup.api

/**
 * Where the access and refresh tokens live, and the only thing the network layer knows about auth.
 *
 * WHY THIS IS AN INTERFACE
 *
 * The generated API interfaces must never mention a token. If they did, every call site would be
 * responsible for attaching one, and the first one anybody forgets is an endpoint that silently
 * behaves as though the user were signed out. Auth is attached once, by an OkHttp interceptor, and
 * this is the only thing that interceptor depends on.
 *
 * It also keeps the real storage out of the network layer's tests: [InMemoryTokenStore] is enough
 * to test injection and refresh behaviour without Android's keystore, an emulator, or a device.
 */
interface TokenStore {
    /** The current access token, or null when nobody is signed in. */
    suspend fun accessToken(): String?

    /** The current refresh token, or null when nobody is signed in. */
    suspend fun refreshToken(): String?

    /** Replace both tokens. Called after sign-in and after a successful refresh. */
    suspend fun save(accessToken: String, refreshToken: String)

    /**
     * Forget both tokens.
     *
     * Called on sign-out, and also when a refresh fails: at that point the session is gone whatever
     * the user does next, and keeping a dead token only produces a second confusing failure.
     */
    suspend fun clear()
}

/**
 * A token store that keeps everything in memory.
 *
 * Used by the tests, and deliberately NOT used by the app: tokens must survive the process being
 * killed, which is what the encrypted store in [EncryptedTokenStore] is for.
 */
class InMemoryTokenStore(
    private var access: String? = null,
    private var refresh: String? = null,
) : TokenStore {
    override suspend fun accessToken(): String? = access

    override suspend fun refreshToken(): String? = refresh

    override suspend fun save(accessToken: String, refreshToken: String) {
        access = accessToken
        refresh = refreshToken
    }

    override suspend fun clear() {
        access = null
        refresh = null
    }
}
